### ISSUE TITLE
[EKS] [request]: Manage IAM identity cluster access with EKS API

### ISSUE THREAD

**Issue by ayosec** [aws/containers-roadmap#185]
**Tell us about your request**

CloudFormation resources to register IAM roles in the [`aws-auth` ConfigMap](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html).

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

A Kubernetes cluster managed by EKS is able to authenticate users with IAM roles. This is very useful to grant access to Lambda functions. However, as described in the [documentation](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html), every IAM role has to be registered manually in a *ConfigMap* with the name `aws-auth`.

For every IAM role we add to the CloudFormation stack, we have to add an entry like this:

```yaml
mapRoles: |
  - rolearn: "arn:aws:iam::11223344:role/stack-FooBarFunction-AABBCCDD"
    username: lambdafoo
    groups:
      - system:masters
  - ...
```

This process is a bit tedious, and it is hard to automate.

It will be much better if those IAM roles can be registered directly in the CloudFormation template. For example, with something like this:

```yaml
LambdaKubeUser:
  Type: AWS::EKS::MapRoles::Entry
  Properties:
    Cluster: !Ref EKSCluster
    RoleArn: !GetAtt FunctionRole.Arn
    UserName: lambdafoo
    Groups:
      - system:masters
```

Thus, CloudFormation will add and remove entries in the ConfigMap as necessary, with no extra manual steps.

Another `AWS::EKS::MapUsers::Entry` can be used to register IAM users in `mapUsers`.

With this addition, we can automate the extra step to register the IAM role of the worker nodes when a new EKS instance is created:

```yaml
NodeInstanceKubeUser:
  Type: AWS::EKS::MapRoles::Entry
  Properties:
    Cluster: !Ref EKSCluster
    RoleArn: !GetAtt NodeInstanceRole.Arn
    UserName: system:node:{{EC2PrivateDNSName}}
    Groups:
      - system:bootstrappers
      - system:nodes
```

**Comment by abelmokadem** [comment]
@ayosec have you created something to automate this as of now? I'm running into this when setting up a cluster using CloudFormation.

**Comment by ayosec** [comment]
>  have you created something to automate this as of now?

Unfortunately, no. I haven't found a reliable way to do it 100% automatic. 

> Do you mind sharing your current approach?

My current approach is to generate the ConfigMap using a template:

1. All relevant ARNs are available in the outputs of the stack.
2. A Ruby script reads those outputs, and fills a template.
3. Finally, the generated YAML is applied with `kubectl apply -f -`.

**Comment by dardanbekteshi** [comment]
Adding this feature on CloudFormation would allow the same feature to be added on [AWS CDK](https://awslabs.github.io/aws-cdk/refs/_aws-cdk_aws-eks.html). This will greatly simplify the process of adding/removing new nodes, for example.

**Comment by willejs** [comment]
I also thought about this. An api to manage the config map for aws-iam-authenticator is interesting, i think would be a bit clunky. I am using terraform to create an eks cluster, and this approach is alot nicer
https://github.com/terraform-aws-modules/terraform-aws-eks/pull/355

**Comment by inductor** [comment]
I'd love this

**Comment by schlomo** [comment]
Anybody from AWS care to comment on this feature request?

**Comment by mikestef9** [comment]
With the release of [Managed Nodes](https://github.com/aws/containers-roadmap/issues/139) with [CloudFormation support](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-eks-nodegroup.html), EKS now automatically handles updating aws-auth config map for joining new nodes to a cluster. 

Does this satisfy the initial use case here, or is there a separate ask to manage adding users to the aws-auth config map via CloudFormation?

**Comment by inductor** [comment]
@mikestef9 I think that https://github.com/aws/containers-roadmap/issues/554 can be one of the similar issues why you would like this kind of options

**Comment by ayosec** [comment]
@mikestef9 

> Does this satisfy the initial use case here, or is there a separate ask to manage adding users to the aws-auth config map via CloudFormation?

My main use case is with Lambda functions.

The managed nodes feature is pretty cool, and very useful for new EKS clusters, but most of our modifications to the `aws-auth` ConfigMap are to add or remove roles for Lambda functions.

**Comment by tnh** [comment]
@mikestef9 It would be useful to then allow people / roles to be able to then run kubectl commands.


Right now, we have a CI deploy role - but we want to allow other saml based users to be able to kubectl 


We do a post cluster creation to kubectl 

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - rolearn: {{ ClusterAdminRoleArn }}
      username: system:node:{{ '{{EC2PrivateDNSName}}' }}
      groups:
        - system:bootstrappers
        - system:nodes
    - rolearn: arn:aws:iam::{{ AccountId }}:role/{{ AdminRoleName }}
      username: admin
      groups:
        - system:masters
    - rolearn: arn:aws:iam::{{ AccountId }}:role/{{ CIRoleName }}
      username: ci
      groups:
        - system:masters
    - rolearn: arn:aws:iam::{{ AccountId }}:role/{{ ViewRoleName }}
      username: view
```
But I'd much rather have this configMAp created by me during the cluster creation

**Comment by nckturner** [comment]
Issues that EKS users have debugging authentication issues (see https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/174 and https://github.com/kubernetes-sigs/aws-iam-authenticator/issues/275) make a strong case for cloudformation management or an "auth-mapping API" (and configurable admin role: https://github.com/aws/containers-roadmap/issues/554)

**Comment by nemo83** [comment]
This ^^, how can we get this implemented? Can anyone from AWS tell us if they gonna support this at CF template level? Or a workaround is needed at eksctl level?

**Comment by inductor** [comment]
@nemo83 AWS team tagged `researching` this issue, so that's the stage of this issue.

**Comment by nicorikken** [comment]
I'm also looking into automating the updates to this configmap from cloudformation. Doing so via lambda seems doable.

My main concern with automation are race-conditions on the contents of the configmap when applying updates as the content has to be parsed. A strategic merge is not possible. **If the configuration would be implemented in one or more (one per entry) CRD's** it would be easier to apply a patch. In that case existing efforts on Kubernetes support for CloudFormation like [kubernetes-resources-provider](https://github.com/aws-quickstart/quickstart-kubernetes-resource-provider/tree/master/apply) can be reused.

**Comment by hellupline** [comment]
I would go a extra mile and ask AWS to create an API to manage aws-auth, with IAM action associated

in case I delete the IAM role/user associated with the cluster creation ( detail: this user/role is not visible after, you have to save this info outside the cluster, or tag the cluster with it. )

and if  I dont add another admin to the cluster, I am now locked out of the cluster, 


for me, this is a major issue, because I use federated auth, users ( and my day-to-day account ) are efemeral... my user can be recreated without warning with another name/ID,



the ideia is: can AWS add an IAM action like ESHttpGet/ESHttpPost ? ( example from ElasticSearch, because is a third party software )

**Comment by mikestef9** [comment]
Hey @hellupline 

We are actually working on exactly that right now, an EKS API to manage IAM users and their permissions to an EKS cluster. This will allow you to manage IAM users via IaC tools like CloudFormation

**Comment by yanivpaz** [comment]
@mikestef9  how it's going to be different compared to https://github.com/aws-quickstart/quickstart-amazon-eks/blob/main/templates/amazon-eks-controlplane.template.yaml#L109

**Comment by markus7811** [comment]
I wonder, why this isn't possible with eks clusters (but with selfhosted k8s clusters on AWS?)

https://github.com/kubernetes-sigs/aws-iam-authenticator#crd-alpha

Even looking at the cdk implementation of auth mapping, it would be simple to get rid of some limitations that exist right now (stack barrier, imported clusters ...)

So if something like this will be implemented (i support this) it would be good, if it won't conflict with the crd's I hope coming soon to eks.

**Comment by gp42** [comment]
> Hey @hellupline
> 
> We are actually working on exactly that right now, an EKS API to manage IAM users and their permissions to an EKS cluster. This will allow you to manage IAM users via IaC tools like CloudFormation

Any news on this issue?

**Comment by vennemp** [comment]
For those wanting to know, this is currently supported via aws cdk - they create a custom resource which invokes a lambda function, and updates the config map.  A little hacky and glad I didn't have to write that code but it works like charm!

**Comment by dany74q** [comment]
@vennemp - Cool ! From what I gather though, that's only possible if it makes sense to assign said lambda the cluster-creation role;
This might be problematic if the clusters were created by a 'plain' user, or by a highly-privileged role, which would be risky to assign.

I think that the 'lock-in' to the cluster creation role is problematic - especially if clusters are created in a non-centralized fashion; it would be awesome to have the ability to control permissions in AWS-alone - along w/ predefined roles one could grant (a-la Kubernetes Viewer / Editor / Owner)

This would also give one the ability to grant said roles to third parties for cluster-auditing, introspection, or just as an auth mechanism - that would work cluster-wide and not require a per-cluster tweak.

**Comment by vennemp** [comment]
@dany74q yes that is what it does and i agree wholeheartedly. It’s definitely a pain. Adding the ability to update the Config mp at launch to include other principals would be the best way  One thing I’ve done is edit the trust relationship of the role to used to create the cluster to include. Trust relationship with another principal - like ec2 or an aws account or another role(you can have more than one!). So I can use the role arn attribute in aws eks update-kubeconfig Command. That way I can just assume that original role.

**Comment by mikestef9** [comment]
@dany74q 

"it would be awesome to have the ability to control permissions in AWS-alone - along w/ predefined roles one could grant (a-la Kubernetes Viewer / Editor / Owner)"

This is exactly the API we are building that I referenced in a previous [comment](https://github.com/aws/containers-roadmap/issues/185#issuecomment-706636321). We'll have a handful of pre-baked polices (Admin, Editor, Viewer) that you can attach through the API at a cluster wide or specific namespace level, without ever having to add anything in RBAC. You can still of course use RBAC if you need to define more fine-grained permissions per IAM entity.

**Comment by tsndqst** [comment]
@mikestef9 is there a separate issue in the roadmap for the API you mentioned or should we follow this issue for updates on that API?

**Comment by mikestef9** [comment]
Let's use this issue, I updated the title

**Comment by manuelcoppotelli** [comment]
The [aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator) has implemented the `IAMIdentityMappings ` CRD.

May be a good idea to follow along with this solution and implement the same!

**Comment by gunzy83** [comment]
Going to drop my $0.02 here and share the use case. We are rebuilding our IAM across all of our accounts with AWS SSO as the entry point to our accounts (console and CLI) so we don't have credentials to distribute anymore. As a part of that, we are trying to follow recommendations for everything we do (least privilege, etc) and Cloudformation service roles are a part of that.

I was talking with our SA contact from AWS about how the role that creates the cluster is the only principal that can perform the update to the config map but if that is a role we pass to Cloudformation when we are automatically deploying our entire stack, we don't want to allow our CI nodes to assume role to that role, update kubeconfig and be able to do everything `system:masters` can do (mind you it can now, we want to do better). 

Our SA pointed me to https://github.com/aws-samples/eks-aws-auth-configmap as a starting point to create a custom resource and allow the lambda to assume the Cloudformation service role. This is better but still not ideal as the service role would have permission to create roles (with a boundary policy), access the KMS key created in Cloudformation, and more; quite a bit more permission than required.

I really hope this will be implemented in the API and eventually added to Cloudformation.

**Comment by adammw** [comment]
@mikestef9 is there further work ongoing on a different API for managing IAM identities, or is the `IAMIdentityMappings` CRD support essentially delivering this API change and there will just be other UI changes to ease the management of these resources?

reason I ask is that is eksctl work on any IAM identity mapping features is stalled until the new API design is announced: https://github.com/weaveworks/eksctl/issues/874

**Comment by mikestef9** [comment]
It's a totally separate EKS API - not related to the IAMIdentityMappings CRD

**Comment by dany74q** [comment]
Hey @mikestef9 ! Wondered if there's an update on this, this is one of the most exiting features to be implemented in EKS, for our use cases at least - appreciating the complexity of implementing it at scale, a rough estimation (months / quarters away) would be amazing.

Thanks a ton !

**Comment by gp42** [comment]
Here is my take on the challenge to manage the auth ConfigMap. Not exactly what is requested here - but may be helpful for some use-cases. The logic is kind of reverse to the one suggested here - the *aws-auth-operator* is synchronising IAM groups to the *aws-auth* configmap.
 
You can find it here: <https://github.com/gp42/aws-auth-operator>

**Comment by matthewcummings** [comment]
@mikestef9 any updates on this?

**Comment by dukeluke16** [comment]
As part of an upcoming Audit, we require the ability to Identify the IAM Role that is associated with the `cluster_creator`.  Whatever solution EKS provides for managing IAM identity cluster access via EKS API needs to explicitly support `cluster_creator`.

**Comment by lucasff** [comment]
I got bitten by this. It took me too long to figure out I needed to access the cluster via a completely different user, even though I have `*` access on AWS. It doesn't make sense for a managed service.

**Comment by savagete2860** [comment]
I have recently started to use the [AWS quick start EKS CloudFormation registry module](https://github.com/aws-quickstart/quickstart-amazon-eks-cluster-resource-provider). Once the third party registry is enabled the resource type `AWSQS::EKS::Cluster` which allows the management the aws-auth config map.

In addition to this, you can use this to add lambda role arns of bootstrap lambdas or whatever you like. It works well but it would be nice if this was a native CloudFormation feature.

**Comment by raizyr** [comment]
This ABSOLUTELY needs to be part of the EKS APIs,  the idea that both AWS and the customer are trying to manage a single ConfigMap in a cluster is ridiculous and completely breaks any concept of a source of truth for what that ConfigMap should be.

**Comment by jagu-sayan** [comment]
With the number of thumbs up, I don't understand why it's not a priority. And it will symplify [AWS CDK for EKS](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_eks-readme.html) implementation (used custom resource for now) !

**Comment by pavlospt** [comment]
I believe it would be a quite good timing to have an official answer from AWS after almost 3 years 😛

**Comment by vennemp** [comment]
@pavlospt AWS: "It's on the 'road-map'"

**Comment by archoversight** [comment]
Having it be an AWS resource would mean we could use IAM privs to control access to the resource, and it avoids the issue we have no where the IAM user/role that created the cluster has god-mode privileges in the cluster, but if that user/role ever goes away and you mess up the `auth-map` you lose access to your EKS cluster.

**Comment by dudicoco** [comment]
> Having it be an AWS resource would mean we could use IAM privs to control access to the resource, and it avoids the issue we have no where the IAM user/role that created the cluster has god-mode privileges in the cluster, but if that user/role ever goes away and you mess up the `auth-map` you lose access to your EKS cluster.

You can actually just recreate the user/role in order to regain access to the EKS cluster.

**Comment by matthewcummings** [comment]
@Phylu as @archoversight mentioned, that module no longer supports managing the aws-auth configmap.

This is issue is ridiculous fwiw. You typically use at least two toolsets with EKS: 1) Terraform, CloudFormation, Pulumi, etc to manage the cluster level resources and resources outside of the cluster and 2) Helm, Kustomize, etc to manage resources _in the cluster_.

This problem lands right in the intersection of those two sets of tools. Cluster access should really just be done seamlessly (to the user) via IAM and the AWS API. It's also crazy that the person who creates the cluster will ALWAYS have admin level access to cluster.

I'm not going to go "full conspiracy" here but this kind of problem makes it clear that people at AWS do NOT use EKS anywhere internally. Which is fine, but if one or two AWS teams had to deal with these problems then they would likely see how awful this is and do something about it.

**Comment by bryantbiggs** [comment]
> > We are creating our EKS clusters using the [terraform-aws-eks module](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/tag/v18.0.0).
> > ...
> > This workflow is – unfortunately – not really feasible in an automated way when creating a new EKS cluster. It would be really helpful, if we can manage the access from the AWS side directly.
> 
> @Phylu You can use Cloud Posse's [terraform-aws-eks-cluster](https://github.com/cloudposse/terraform-aws-eks-cluster) module instead. If you can run your Terraform in an environment where the official `aws` CLI is installed, then Cloud Posse's module will take care of the `auth-map` reasonably well. (It works without the `aws` CLI as well, but not as reliably.) Note that the Cloud Posse module still has edge cases where it requires manual intervention (for example when the initial `destroy` only partly succeeds, a subsequent `destroy` might fail due to failing authentication), and the better solution is to implement the API as requested in this issue. The Cloud Posse solution is nevertheless IMHO the best option available at this time for a fully automated solution.

@aidanmelen has provided a module for managing the `aws-auth` configmap that works quite well to fill this gap until a native solution is provided https://github.com/aidanmelen/terraform-aws-eks-auth

**Comment by matthewcummings** [comment]
To be clear, this functionality should be native to the AWS API. Custom terraform modules (that leverage the K8s provider)and custom Cfn resources are all nice and good but this should be an API level operation. Anything short of that is a compromise.

I've been using AWS for years, ECS (Fargate and EC2), Lambda, EKS and of course tons of other services but this is a "leaky abstraction" at best. There should be API level support for assigning cluster IAM permissions. Bonafide API support makes it possible to do this in CloudFormation and Terraform (with just the AWS provider) and Pulumi and everything else. I'm not heartless but I also expect a lot from AWS.

**Comment by aidanmelen** [comment]
> To be clear, this functionality should be native to the AWS API. Custom terraform modules (that leverage the K8s provider)and custom Cfn resources are all nice and good but this should be an API level operation. Anything short of that is a compromise.
> 
> 
> 
> I've been using AWS for years, ECS (Fargate and EC2), Lambda, EKS and of course tons of other services but this is a "leaky abstraction" at best. There should be API level support for assigning cluster IAM permissions. Bonafide API support makes it possible to do this in CloudFormation and Terraform (with just the AWS provider) and Pulumi and everything else. I'm not heartless but I also expect a lot from AWS. 
> 
>
Yes. AWS needs to make an API for this.

**Comment by andre-lx** [comment]
Hey guys.

I'm searching everywhere and can't find the answer, right now, how can we accomplish this? 

So, I'm creating a private cluster through CloudFormation and want to give access to a role used by an ECS Task. The security groups are configured and now the ECS task can access the cluster (network), but I'm receiving the "error: You must be logged in to the server (Unauthorized)" as expected. 

There are any chance of setting this role to the "aws-auth" configmap in order to make this connection possible? 

Thanks

**Comment by matthewcummings** [comment]
@andre-lx your best best is to switch to Terraform or Pulumi and use the Kubernetes provider. CloudFormation only natively works with AWS resources i.e. not Kubernetes.

**Comment by vennemp** [comment]
https://aws.amazon.com/blogs/containers/bootstrapping-clusters-with-eks-blueprints/

Somewhat of a step in the right direction.

**Comment by vbala81** [comment]
@vennemp  Thanks for sharing the link. This is still only for CDK and Terraform options. I am also trying to automate the eks authorization similar to others in this group and I can only use CFT due to project constraints. I will also raise this up with my TAM in AWS to see if they can expedite this request

**Comment by andre-lx** [comment]
Not the optimal solution, but, from our understanding the only way of achieving this in CloudFormation, currently, is using the AWSQS::EKS::Cluster

https://github.com/aws-quickstart/quickstart-amazon-eks-cluster-resource-provider

The configuration is identical to the official EKS Cluster, and it's working properly. 

Unfortunately, this means that we need to add an extra step setting the IAM Role and activating the extension.

**Comment by MarcusRosen-Rio** [comment]
> Hey guys.
> 
> I'm searching everywhere and can't find the answer, right now, how can we accomplish this?
> 
> So, I'm creating a private cluster through CloudFormation and want to give access to a role used by an ECS Task. The security groups are configured and now the ECS task can access the cluster (network), but I'm receiving the "error: You must be logged in to the server (Unauthorized)" as expected.
> 
> There are any chance of setting this role to the "aws-auth" configmap in order to make this connection possible?
> 
> Thanks

You can use the custom AWS QuickStart Cloudformation type, if you want to do this completely through CloudFormation.
We use it and it works ok but it's a workaround..
https://github.com/aws-quickstart/quickstart-amazon-eks-cluster-resource-provider/blob/main/README.md

**Comment by aavileli** [comment]
AWS needs to make an API for configuring aws-auth to configure access to cluster to make provisioning tools like terraform handle it correctly when re-applying config changes

**Comment by musayev-io** [comment]
+1 -- this should be natively supported, not a jerry-rig through custom resources.

**Comment by sharma-abhishek-it** [comment]
+1

**Comment by igorGevaerd** [comment]
This really would help automate things up, removing the manual changes!

**Comment by ManuelMueller1st** [comment]
Customer Obsession. Sure.

**Comment by MartinEmrich** [comment]
The discussion so far seems to mostly revolve around IaC and configuration, where numerous workarounds or 3rd-party-plugins seem to be sprouting everywhere.

But the current implementation seems to be actually quite fragile, even dangerous:

The User (exactly: The IAM ARN) who initially created the cluster always has full privileges on the cluster. This is hidden somewhere deep inside AWS, I read many times that it cannot be viewed or changed.
Even a small typo in the `aws-auth` ConfigMap makes it ineffective. (The fact that `mapUsers` and `mapRoles` are inline multiline 
 strings does not help here).

So there are easy scenarios where you can lock yourself out of your own EKS cluster:

Maybe the IAM user who created it moved on to another team, and can no longer help. (in that case, you cannot even lock it out easily). Or the cluster is so old, that no one even knows who created it in the first place, so you cannot recreate the very IAM user who did it.
Or you moved from IAM users to AWS IAM Identity Center (SSO), which is best practice anyways.... And doing AWS SSO with EKS (PermissionSets etc.) and the aws-auth ConfigMap is IMHO kludgy as well....

If you now damage your `aws-auth` ConfigMap accidentally, there's no way back in. Furthermore, all the node group roles referenced no longer work either, leaving the cluster in a broken state if e.g. Cluster Autoscaler is scaling node groups.

IMHO the only way is to get rid of that "initial creator" concept entirely, and provide a validated (meaning bullet-proof) alternative to the aws-auth ConfigMap, held outside of the actual Kubernetes cluster.

What's your take on this?

**Comment by matthewcummings** [comment]
@MartinEmrich not only is it dangerous, it's also counter-intuitive. The default cluster admin permission are unique in the AWS ecosystem. I would bet that most/all of the people following this issue have wasted hours the first time we encountered this behavior, I definitely did.

**Comment by mikestef9** [comment]
@MartinEmrich appreciate the feedback and concern. What you are describing is exactly what we are building. An EKS API replacement for the aws-auth config map. With this API, you will also be able to control whether or not the cluster creator has admin privileges.

**Comment by blakepettersson** [comment]
Is there a timeline (months, years, ...) for this feature?

**Comment by ayosec** [comment]
@mikestef9 Thanks for the update!

Can you share any estimates of when it will be available?

Will be possible to use that API entirely from CloudFormation?

**Comment by jaredk-beta** [comment]
@MartinEmrich You can use an IAM role to create the cluster, which is a much better approach than using an IAM user. Any admin that can assume the role can still manage the cluster if the aws-auth configmap gets misconfigured.

An API to replace the brittle aws-auth configmap would be much more useful though as others have mentioned. This would make deploying EKS much simpler from an IaC perspective.

**Comment by mikestef9** [comment]
Can't give anything too specific in this forum - but this is in active development. Will move to coming soon once I feel it aligns the roadmap categories defined here https://github.com/aws/containers-roadmap#faqs  And yes, this will be an AWS API - and it will include CloudFormation support.

**Comment by MartinEmrich** [comment]
@jaredk-beta thanks, that's probably what we will do at least for new clusters until the proper solution is available.

**Comment by m1mohamad** [comment]
I made a mistake and updated my cluster with the following configmap while I was trying to test why nodes previously arent getting READY , so I lost access to the cluster, then I used the same role below to access the cluster again but I lost my system:masters permissions ! I can't get do kubectl get configmaps , is there a way to regain access  ? 
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - rolearn: 'arn:aws:iam::xxxxxxx:role/testyy-NodeInstanceRole-1FQVVVZPS0TDP'
      username: system:node:{{EC2PrivateDNSName}}
      groups:
        - system:bootstrappers
        - system:nodes

```

**Comment by aservedio-swi** [comment]
> I made a mistake and updated my cluster with the following configmap while I was trying to test why nodes previously arent getting READY , so I lost access to the cluster, then I used the same role below to access the cluster again but I lost my system:masters permissions ! I can't get do kubectl get configmaps , is there a way to regain access ?
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: aws-auth
>   namespace: kube-system
> data:
>   mapRoles: |
>     - rolearn: 'arn:aws:iam::xxxxxxx:role/testyy-NodeInstanceRole-1FQVVVZPS0TDP'
>       username: system:node:{{EC2PrivateDNSName}}
>       groups:
>         - system:bootstrappers
>         - system:nodes
> ```

@m1mohamad - the user that you used to create the cluster does not appear in this configmap and should still have access. If you are able to use credentials from them, I think you should be able to fix your configmap.

**Comment by LeoSpyke** [comment]
How is it possible that a crucial feature like this it is still in "Open" state after 3 years?

**Comment by mikestef9** [comment]
This is a major new feature, so we are being careful with testing and rollout, but this is coming soon. This feature will address the following open issues #554 #922 #1209 #1278 #573 #795 #822 #727 #926. 

At launch, it will not make the integration with SSO any simpler #474, but we will share details on a follow on enhancement  that will directly integrate with AWS IAM Identity Center Permission sets.

**Comment by ClaytonOlleyNutrien** [comment]
I'm in dire need of this functionality as well. Managing EKS via Terraform has been quite a nightmare for me with various issues with aws-auth cropping up constantly with no apparent long term fixes.

**Comment by MartinEmrich** [comment]
@mikestef9 I understand that doing this "right" (no breaking changes, no possibilities to shoot in the own foot, no security holes) is a big challenge and it takes time.

But until the Big Bang: Could you have the AWS team at least implement a simple API that allows viewing and replacing the currently hidden IAM principal (that created the cluster in the first place and always has full admin access regardless of the *aws-auth* ConfigMap)?

I bet that would be comparatively simple, and would save a lot of bacon (or veggie equivalents) around here...

Thanks so much!

**Comment by mikestef9** [comment]
You are right that request isn't directly related to this feature, but we are coupling support for that request together in the same launch and set of deployments, so at this point there isn't any way to decouple them. 

I know how many of you are waiting on this feature, so I'll share a sneak preview of the user guide for this particular request to control cluster creation permissions.

-----

The following command will create a cluster without any IAM identity having admin access to the cluster. 
```
aws eks create-cluster \
  --name my-cluster \
  --role-arn arn:aws:iam::012345678910:role/myClusterRole \
  --resources-vpc-config subnetIds=subnet-6782e71e,subnet-e7e761ac
  --no-bootstrap-cluster-creator-admin-permissions
```

After cluster creation, you can create a standard access entry with whatever desired IAM identity as a cluster admin. 

For existing clusters, after being updated to an EKS platform version that supports access entry and opting in to access entry APIs, the cluster creator IAM identity will now be returned as an existing access entry. You can choose to delete this access entry if desired.
```
aws eks list-access-entries --cluster-name my-existing-cluster 
{
  "accessEntries": [ 
     "arn:aws:iam::012345678910:role/EKSClusterCreatorCICDRole",
   ]
}

aws eks delete-access-entry --cluster-name my-existing-cluster \
  --principal-arn arn:aws:iam::012345678910:role/EKSClusterCreatorCICDRole
```

If you don’t set the admin bootstrap parameter (or explicitly set it to True), EKS will automatically create a cluster admin access entry on your behalf. The bootstrap parameter can be set independent of cluster authentication mode (meaning you can set this parameter even with authentication mode set to CONFIG_MAP). However, if you have your cluster authentication mode set to CONFIG_MAP, the cluster creator Kubernetes admin role will effectively remain hidden, because the access entry APIs are disabled. 

Note: The value that you set for bootstrapClusterCreatorAdminPermissions is not returned in the EKS DescribeCluster API. Further changes to access control post cluster creation will always be performed with access entry APIs.

**Comment by dvins** [comment]
> This is exactly the API we are building that I referenced in a previous [comment](https://github.com/aws/containers-roadmap/issues/185#issuecomment-706636321). We'll have a handful of pre-baked polices (Admin, Editor, Viewer) that you can attach through the API at a cluster wide or specific namespace level, without ever having to add anything in RBAC. You can still of course use RBAC if you need to define more fine-grained permissions per IAM entity.

We're approaching two and a half years since this comment. Is there a timeframe as for when this is going to finally be available? The recent comments that the API support is tightly coupled with a whole bunch of other capabilities is disheartening given how long this issue has been outstanding.

**Comment by dvins** [comment]
> This is a major new feature, so we are being careful with testing and rollout, but this is coming soon. This feature will address the following open issues #554 #922 #1209 #1278 #573 #795 #822 #727 #926.
> 
> At launch, it will not make the integration with SSO any simpler #474, but we will share details on a follow on enhancement that will directly integrate with AWS IAM Identity Center Permission sets.

Also, regarding SSO, how closely (or not) will this support follow?

**Comment by ceresdarkmatter** [comment]
> > This is exactly the API we are building that I referenced in a previous [comment](https://github.com/aws/containers-roadmap/issues/185#issuecomment-706636321). We'll have a handful of pre-baked polices (Admin, Editor, Viewer) that you can attach through the API at a cluster wide or specific namespace level, without ever having to add anything in RBAC. You can still of course use RBAC if you need to define more fine-grained permissions per IAM entity.
> 
> We're approaching two and a half years since this comment. Is there a timeframe as for when this is going to finally be available? The recent comments that the API support is tightly coupled with a whole bunch of other capabilities is disheartening given how long this issue has been outstanding.

Knowing the history of this garbage fire of a company - probably this will never see the light of day

Let's see what is the new excuse for this situation from these excuse of drones... ah no, what was the accepted word for them? Ah yes, "employees" I think

**Comment by bryantbiggs** [comment]
@ceresdarkmatter lets please be civilized and avoid negative/derogatory comments

**Comment by ceresdarkmatter** [comment]
> @ceresdarkmatter lets please be civilized and avoid negative/derogatory comments

Bryant, I am trying but see the results of a multy-year civilised conversation. I mean - you have to get the attention about something in a way. 
I am loosing my patience with updates every few months that just move the goal posts.
Those only indicate that probably besides the dog and pony show AWS puts in marketing materials they don't really care about the actual needs of paying users - and this situation is not the first time that I have seen this exact scenario unfolding with them.
I am seriously asking what are professional ways to solve this systemic problem besides moving to another cloud provider

**Comment by jking916** [comment]
@ceresdarkmatter As a paying customer, if you don't like what AWS is doing then take your business to another cloud provider that you feel cares more about their customers. But insulting the AWS employees in here because they don't share your viewpoint isn't appropriate or helpful.

**Comment by ceresdarkmatter** [comment]
> @ceresdarkmatter As a paying customer, if you don't like what AWS is doing then take your business to another cloud provider that you feel cares more about their customers. But insulting the AWS employees in here because they don't share your viewpoint isn't appropriate or helpful.

@jking916  Those AWS employees are in charge of getting this issue fixed, do you realise that? It is 100% on them to get their shit together and make it happen.
Them not doing that are an insult to everyone that explicitly requested this since before the Pandemic. I am not cutting them any slack and the fact that absolutely no one is trying to keep them accountable is absolutely mindblowing.

Having said that I see that I have no one to discuss with and you guys are part of the problem.
This is the last reply I am going to post... unbelievable

**Comment by gazal-k** [comment]
I think I can empathize with AWS on this one. Technology teams working on popular product platforms would like to work on a whole bunch of things, but the reality is that it is impossible to prioritise all of them. The teams around EKS have been consistently improving those platforms with new features. Would I love to see this improvement to EKS <> IAM? Absolutely, but we've been very happy with EKS since launch (back in 2019, I believe).

**Comment by nnsense** [comment]
The way @ceresdarkmatter is phrasing his message isn't the most calm, OK, but.. have you ever found yourself with a prod cluster locked up because of this issue? with tens devs unable to access the cluster and you eventually having to fix it manually because terraform keeps showing "Unauthorized"? It happened to me, many times over these years, bear in mind this issue has been opened in 2019 but I've had this problem before that. Sure we need to be polite, but I can understand the frustration, and I hope the feature fixing these problems will be published as soon as possible.

**Comment by gazal-k** [comment]
We create and manage all infra using CI / CD. The cluster is created using the pipeline's IAM Role. That way, we can never lock ourselves out. If you don't have such a pipeline, you can still use a Role to create clusters (for now) to prevent locking yourself out. The locking out part wasn't our primary motivator for this mechanism, just that we wanted to ensure there was some rigour around how we managed infra.

**Comment by vennemp** [comment]
@gazal-k is right roles are the best for creation and you can even give them multiple trust relationships - though I know some corporate policies do not permit this.  One thing I tested, if you create a cluster with an IAM user and that user gets deleted - if you re-create an iam user with the same name, you should still be able to access the cluster with the new IAM user since the principal is still the same.

**Comment by RothAndrew** [comment]
@mikestef9 any update?

**Comment by ab77** [comment]
Going to leave this here, in case it helps someone. We use the following workaround using clunky CFN [custom resources](https://github.com/belodetek/cfn-generic-custom-resource). Implementation detail, broadly:

* push the Lambda image from public https://gallery.ecr.aws/k4v3r5y6/cfn-generic-provider to your ECR private repo
* deploy [custom resource Lambda](https://github.com/belodetek/cfn-generic-custom-resource) into your account/region
* add a custom resource to your EKS stack(s)

... (e.g.) custom resource function (one per account/region)

```
  Function:
    Type: AWS::Lambda::Function
    Properties:
      FunctionName: custom-resource
      Code:
        ImageUri: !Ref EcrRepositoryUri
      PackageType: Image
      # both architectures are in the manifest, pick one
      Architectures:
        - !Ref EcrRepositoryImageArch
      Role: !GetAtt Role.Arn
      Timeout: 300
      Environment:
        Variables:
          VERBOSE: 1
```


... (e.g.) custom resource usage example

```
  PatchConfigMapAwsAuth:
    Type: 'Custom::PatchConfigMap'
    Version: 1.0
    Properties:
      # assume cluster creation role to perform AWS/EKS operations (e.g. GitHub OIDC role used by CI)
      RoleArn: !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/github-oidc"
      # reference to the Lambda created in the first step
      ServiceToken: !Sub 'arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:custom-resource'
      AgentService: eks
      AgentType: custom
      AgentCreateMethod: update_aws_auth_configmap
      AgentCreateArgs:
        ClusterName: !Sub "${ClusterName}"
        AgentRegion: !Sub "${AWS::Region}"
        RoleArn: !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/admin"
        Username: balena:admin
        Groups:
          - admin
          - cluster-admin
          - system:masters
      AgentUpdateMethod: update_aws_auth_configmap
      AgentUpdateArgs:
        ClusterName: !Sub "${ClusterName}"
        AgentRegion: !Sub "${AWS::Region}"
        RoleArn: !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/admin"
        Username: balena:admin
        Groups:
          - admin
          - cluster-admin
          - system:masters
      AgentDeleteMethod: delete_aws_auth_configmap
      AgentDeleteArgs:
        ClusterName: !Sub "${ClusterName}"
        AgentRegion: !Sub "${AWS::Region}"
        RoleArn: !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/admin"
```

The example custom resource will call [this](https://github.com/belodetek/cfn-generic-custom-resource/blob/master/generic_provider/eks.py#L114-L134) code and patch `aws-auth` config map.

The code may expose your AWS session keys/k8s tokens in the CloudWatch Lambda logs.

`aws-auth` ConfigMap must already exist (i.e after the first nodegroup is created).

**Comment by jagu-sayan** [comment]
Or if you use CDK : https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.AwsCustomResource.html

**Comment by ab77** [comment]
> Or if you use CDK : https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.AwsCustomResource.html

Yeah, that's an option if the resources are created with CDK. We currently don't need/use OO IAC, so we still write plain CFN templates YML. Using a custom resource seems like a reasonable approach, which is basically what aws-cdk does under the hood anyway.

**Comment by rrsela** [comment]
Isn't this request fulfilled with the latest API change regarding AccessEntry? See [this](https://aws.amazon.com/blogs/containers/a-deep-dive-into-simplified-amazon-eks-access-management-controls/) blog post for details..

**Comment by bryantbiggs** [comment]
Yes!

**Comment by matthewcummings** [comment]
NICE

**Comment by mikestef9** [comment]
EKS now supports a simplified method for managing IAM authentication and authorization. 
- [What's new](https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-eks-controls-iam-cluster-access-management/)
- [Blog](https://aws.amazon.com/blogs/containers/a-deep-dive-into-simplified-amazon-eks-access-management-controls/)
- [Docs](https://docs.aws.amazon.com/eks/latest/userguide/access-entries.html)

**Comment by mikestef9** [comment]
## More details on the feature

**Enable Access Management APIs**

EKS access entry APIs require you to opt in before use. This can be done during cluster creation or updated on existing clusters. You must first ensure your cluster is on a supported [platform version](https://docs.aws.amazon.com/eks/latest/userguide/platform-versions.html).

**Supported platform versions**

| Kubernetes version | EKS Platform version |
| --- | ----------- |
| 1.28 | eks.6 |
| 1.27 | eks.10 |
| 1.26 | eks.11 |
| 1.25 | eks.12 |
| 1.24 | eks.15 |
| 1.23 | eks.17 |

```
aws eks create-cluster \
   --name my-cluster \
   --role-arn arn **:** aws **:** iam **::** 012345678910 **:** role/myClusterRole \
   --resources-vpc-config subnetIds=subnet-6782e71e **,** subnet-e7e761ac
   --access-config authenticationMode=API
```

Existing cluster
```
aws eks update-cluster-config \
   --name my-cluster \
   --access-config authenticationMode=API
```

EKS supports the following three values for authenticationMode:

- `CONFIG_MAP`: The cluster will source authenticated IAM principals only from the aws-auth ConﬁgMap.
- `API_AND_CONFIG_MAP`: The cluster will source authenticated IAM principals from both EKS access entry APIs and the aws-auth ConﬁgMap, with the EKS access entry APIs taking priority.
- `API`: The cluster will source authenticated IAM principals only from EKS access entry APIs.

In a to be determined future Kubernetes version, EKS will stop supporting the aws-auth ConfigMap as an authentication source, and you will be blocked from upgrading if you haven't switched your cluster authenticationMode to opt in to access entry APIs. Users are strongly encouraged to migrate to access entry APIs as soon as possible.

**Note**: The default selected option at launch for new clusters in the EKS Console is `API_AND_CONFIG_MAP`.

Switching authentication modes on an existing cluster is a one-way operation. You can switch from `CONFIG_MAP` to `API_AND_CONFIG_MAP`. You can then switch from `API_AND_CONFIG_MAP` to `API`. You cannot revert these operations in the opposite direction. Meaning you cannot switch back to `CONFIG_MAP` or `API_AND_CONFIG_MAP` from `API`. And you cannot switch back to `CONFIG_MAP` from `API_AND_CONFIG_MAP`.

`eks:authenticationMode` is a supported IAM condition key, so you can write IAM policies that enforce clusters be created with a certain authentication mode.

### **Managing/Modifying Cluster Admin**

**Control over which identity has cluster admin permissions when creating a cluster. The following command will create a cluster without any IAM identity having Kubernetes cluster admin access to the cluster.**

```
aws eks create-cluster \
   --name my-cluster \
   --role-arn arn **:** aws **:** iam **::** 012345678910 **:** role/myClusterRole \
   --resources-vpc-config subnetIds=subnet-6782e71e **,** subnet-e7e761ac
   --access-config authenticationMode=API
   --no-bootstrap-cluster-creator-admin-permissions
```
**Important** : bootstrapClusterCreatorAdminPermissions can only be set to False if the cluster authenticationMode is set to either `API` or `API_AND_CONFIG_MAP`.

After cluster creation (or during, as you can create access entries while a cluster is in CREATING state), you can create a standard access entry with whatever desired IAM identity as a cluster admin.

For existing clusters with authenticationMode set to `CONFIG_MAP`, after being updated to an EKS platform version that supports access entry and changing authenticationMode to use access entry APIs, the original IAM identity that created the cluster will be returned as an existing access entry. You can choose to delete this access entry if desired.
```
aws eks list-access-entries --cluster-name my-existing-cluster

{
  "accessEntries": [
    "arn:aws:iam::012345678910:role/EKSClusterCreatorCICDRole",
  ]
}
```

```
aws eks delete-access-entry --cluster-name my-existing-cluster \
   --principal-arn arn:aws:iam::012345678910:role/EKSClusterCreatorCICDRole
```

GitHub issues resolved:

- Add AdminRole option at cluster creation [#554](https://github.com/aws/containers-roadmap/issues/554)
- Add an ability to view and update the IAM entity user or role that is automatically granted system:masters permissions in the cluster's RBAC configuration [#922](https://github.com/aws/containers-roadmap/issues/922)

If you don't set the admin bootstrap parameter (or explicitly set it to True), EKS will automatically create an access entry with Kubernetes cluster administrator permissions on your behalf.

**Note**: The value that you set for bootstrapClusterCreatorAdminPermissions on cluster creation is not returned in the response of subsequent EKS DescribeCluster API calls. This is because the value of that field post cluster creation may not be accurate. Further changes to access control post cluster creation will always be performed with access entry APIs. The ListAccessEntries API is the source of truth for cluster access post cluster creation.

**Recover cluster admin access**

Previously if you made a typo in the aws-auth ConfigMap you could lock yourself out of the cluster. Or you could lose access to cluster if you accidentally deleted cluster creator IAM identity and didn't remember the IAM identity to re-create it. Now, as long as an IAM identity has permission to create access entries, you can always recover Kubernetes API access to your cluster.
```
aws eks create-access-entry --cluster-name my-cluster  \
   --principal-arn arn:aws:iam::012345678910:role/Admin

 aws eks associate-access-policy --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/Admin \
   --policy-arn arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy \
   --access-scope '{"type": "cluster"}'
```

GitHub issues resolved:

- Allow customer to rollback aws-auth configmap when update the configuration wrong (Solving the goal of the request) [#1209](https://github.com/aws/containers-roadmap/issues/1209)

### **View Kubernetes resources from EKS Console**

A common challenge is EKS Console users (who didn't create the cluster and have Kubernetes cluster admin permissions by default) struggling figure out why they can't view Kubernetes resources in the Console. Previously, a separate IAM principal with cluster access would have to add an additional aws-auth ConfigMap entry for that Console user's IAM principal. Now, a pop up will appear for IAM principals using the EKS Console that takes them to the create access entry screen with their IAM principal pre-popluated. As the long as the Console user has IAM permissions to call EKS access entry APIs, they can give themselves Kubernetes permissions from the Console itself without needing create aws-auth ConfigMap entries anymore.

**Note:** The Console user's IAM identity still first requires the permission `eks:AccessKubernetesApi` to view resources in the console, irrespective of if they have Kubernetes level permissions granted through access entry APIs.

GitHub issues resolved:

- Manage aws-auth ConfigMap in AWS Console (access entries don't actually interact with aws-auth ConfigMap, but this solves the goal of the request) [#1278](https://github.com/aws/containers-roadmap/issues/1278)

### **Authentication**

The following examples show to authenticate IAM principals to an EKS cluster, as the preferred alternative to the current experience of using the aws-auth ConfigMap. If no EKS access policies are attached to the principal (as shown in authorization examples below), no Kubernetes permissions are granted, but the principal will be able to successfully authenticate to the cluster, and more fine-grained authorization can be implemented using RBAC.

Username is not a required input when creating an access entry. If not specified, the username value passed to Kubernetes for authentication decisions will map to the value returned from the IAM STS [get-caller-identity API](https://docs.aws.amazon.com/cli/latest/reference/sts/get-caller-identity.html). For IAM users, this is the original IAM user ARN. For IAM roles, this is the assumed role ARN with session name, of the format `arn:aws:sts::012345678910:assumed-role/<role-name-without-path>/{{SessionName}}`. This value can change depending on the session name, so for use with RBAC, you will want to set Kubernetes groups on the access entry to reference in Kubernetes role bindings.

We recommend leaving custom username blank, as the generated username includes the session name, which will appear in CloudTrail logs as an added security benefit. If using a custom username with IAM roles, it is strongly encouraged to set the `{{SessionName}}` macro. You can specify the following macros in custom usernames `{{AccessKeyID}}`, `{{AccountID}}`, `{{SessionName}}`. Custom usernames can't start with `system:`, `eks:`, `aws:`, `amazon:` , or `iam:`.

**Note**: IAM principals referenced in an access entry must already exist in IAM.

**Authenticate an IAM user to a cluster**
```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:user/MyUser
```
**Note**: The Kubernetes username used for further authorization RBAC in this case will be the same as the input principal ARN `arn:aws:iam::012345678910:user/MyUser`. 

**Note**: If you change your IAM username, that user will no longer have access to the cluster. You will need to recreate an access entry with the updated IAM user ARN. 

**Note**: [IAM best practices](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#bp-users-federation-idp) recommends using roles with temporary credentials, rather than users with long-term credentials.

**Authenticate an IAM role to a cluster and specify Kubernetes groups**

```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole \
   --kubernetes-groups dev test
```

**Note**: The Kubernetes username generated in this case will be `arn:aws:sts::012345678910:assumed-role/MyRole/<session-name>`.

**Authenticate an IAM role to a cluster with a custom username**

```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole \
   --username my-username
```

**Note:** No groups are specified here, because the static username of my-username can be referenced in Kubernetes RBAC policies. However, this is not recommended, as the assumed role session name will be lost from Kubernetes audit logs.

**Authenticate an IAM role to a cluster with a custom username using macros**

```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole \
   --username my-username:{{AccountID}}:{{SessionName}}
   --kubernetes-groups dev test
```

**Note**: The macros will be expanded when username reaches Kubernetes for further RBAC authorization decision.

**Note**: As illustrated in this example, when using the `{{SessionName}}` macro, a : must be part of the custom username at some point before the macro (does not have to directly precede as in example above).

**Update the username and/or groups of an existing access entry**

```
aws eks update-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole \
   --username my-other-username
   --kubernetes-groups my-other-group
```

**Note**: Be careful performing this operation if you already have existing RBAC resources that reference previous username/groups. RBAC will also have to be updated to support the new username/groups.

**Note**: Groups are declarative. Enter all desired groups with the update call.

**Creating an access entry with a role ARN containing a path**

Authentication decision for EKS access entries are matched against the IAM [principal ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html) of an IAM identity, rather than a string match against input value of the IAM identity and returned assumed role response from sts get-caller-identity (as is the case with entries in the aws-auth ConfigMap).

When creating an EKS access entry, enter the fully qualified IAM role ARN like below (do not strip off the path like you need to do in the aws-auth ConfigMap today)

```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/teams/tooling/devToolingOperator
```

**Note:** The username in returned access entry will not include the path: `arn:aws:sts::012345678910:assumed-role/devToolingOperator/{{SessionName}}`.

**Authenticate an IAM principal from a separate account**

From account 012345678910 run the following

```
aws eks create-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::567890123456:role/myOtherAccountRole
```

An identity assuming the myOtherAccountRole role in account 567890123456 is now able to authenticate to the my-cluster cluster in account 012345678910.

**Eventual Consistency**

Similar to AWS IAM [behavior](https://docs.aws.amazon.com/IAM/latest/UserGuide/troubleshoot_general.html#troubleshoot_general_eventual-consistency), EKS access entries are eventually consistent, and may take several seconds to be effective after the initial API call returns successfully. You must design your applications to account for these potential delays. We recommend that you do not include access entry create/updates in the critical, high-availability code paths of your application. Instead, make changes in a separate initialization or setup routine that you run less frequently. Also, be sure to verify that the changes have been propagated before production workflows depend on them.

GitHub issues resolved:

- Manage IAM identity cluster access with EKS API [#185](https://github.com/aws/containers-roadmap/issues/185)
- Federated roles containing paths don't work properly with EKS [#573](https://github.com/aws/containers-roadmap/issues/573)
- Privilege Escalation using aws-auth configmap [#795](https://github.com/aws/containers-roadmap/issues/795)
- Add aws-iam-authenticator as a CRD (again, this feature does not interact with the aws-auth config map, but solves the goal of the request) [#550](https://github.com/aws/containers-roadmap/issues/550)

### **Authorization**

The Kubernetes API server supports an ordered list of authorization modules, allowing modules to approve, deny, or pass on a particular action. EKS clusters now include an additional authorizer separate from RBAC, and you can optionally attach EKS defined access policies to give Kubernetes authorization permissions to the IAM principal authenticated with an access entry. Authorization can be done on a namespace or cluster scope. You can specify up to 25 namespace selectors, with optional wildcard suffix matching, to restrict that principals access to specific namespaces.

In EKS clusters, RBAC policies are currently evaluated prior to EKS access policies, so any Kubernetes Role Based Access Control (RBAC) ClusterRoleBinding or RoleBinding that specifies an IAM principal will be evaluated for an allow authorization decision before EKS Access Policies. This may change in the future, although in practice the order today does not matter, because neither RBAC nor EKS access policies support deny actions.

EKS predefines several managed access policies that mirror the default [Kubernetes user facing roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles), including cluster-admin, admin, edit, and, view. To view the list of all policies you can run the following command

`aws eks list-access-policies`

The examples below assume you have already created the access entry to authenticate the IAM principal to the cluster.

**Give read only permissions for an IAM principal to an entire cluster**
```
aws eks associate-access-policy --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole \
   --policy-arn arn:aws:eks::aws:cluster-access-policy/AmazonEKSViewPolicy
   --access-scope '{"type": "cluster"}'
```

**Give editor permissions for an IAM principal to specific namespaces**
```
aws eks associate-access-policy --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole2 \
   --policy-arn arn:aws:eks::aws:cluster-access-policy/AmazonEKSEditPolicy
   --access-scope '{"type": "namespace", "namespaces": ["app-a", "app-b"]}'
```

**Give admin permissions to an IAM identity to access multiple namespaces with a wildcard selector**
```
aws eks associate-access-policy --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole3 \
   --policy-arn arn:aws:eks::aws:cluster-access-policy/AmazonEKSAdminPolicy
   --access-scope '{"type": "namespace", "namespaces": ["dev-\*"]}'
```
**Update list of allowed namespaces for an existing access entry**
```
aws eks associate-access-policy --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole2 \
   --policy-arn arn:aws:eks::aws:cluster-access-policy/AmazonEKSEditPolicy
   --access-scope '{"type": "namespace", "namespaces": ["app-a", "app-c"]}'
```
### **Node Authentication and Authorization**

Access entries have an optional input type field for use with authorizing worker node roles to join a cluster. The following types are supported by the access entry API:

- `EC2_LINUX`
- `EC2_WINDOWS`
- `FARGATE_LINUX`
- `STANDARD`

`STANDARD` is the default type and will be set as the default if not passed on access entry creation. This is the default experience for non-node role access entries. Similar to with the aws-auth ConfigMap today, there is no need to manually create access entries for IAM roles used with managed node groups and Fargate. MNG and Fargate will continue to manage node entries on your behalf, irrespective of authenticationMode. When you switch an existing cluster authenticationMode from `CONFIG_MAP` to `API_AND_CONFIG_MAP` entries in the aws-auth ConfigMap managed by MNG and Fargate will be automatically migrated to node type access entries. Once your cluster is migrated to an access entry supporting authentication mode, MNG and Fargate will only manage entries using access entry APIs, not the aws-auth ConfigMap.

**Note**: aws-auth ConfigMap entries not associated with MNG or Fargate are not automatically migrated when switching authentication modes. These entries must be migrated manually.

**Note**: In practice, you will never need to manually use the `FARGATE_LINUX` type when creating an access entry, because that entry would always be created on your behalf by the EKS/Fargate service. If you do pre-create the access entry, EKS/Fargate will simply skip trying to create the entry itself.

The examples below would be required if using self-managed node groups or Karpenter.
```
aws eks create-access-entry --cluster-name my-cluster  \
   --principal-arn arn:aws:iam::012345678910:role/MyLinuxNodeRole \
   --type EC2_LINUX

aws eks create-access-entry --cluster-name my-cluster  \
   --principal-arn arn:aws:iam::012345678910:role/MyWindowsNodeRole \
   --type EC2_WINDOWS
```

**Note**: It is required to use separate IAM roles for the `EC2_LINUX`, `EC2_WINDOWS`, and `FARGATE_LINUX` types.

**Note**: You cannot attach EKS access policies to access entries with type of `EC2_LINUX`, `EC2_WINDOWS`, or `FARGATE_LINUX`. EKS automatically handles the permissions required for worker nodes with those access entries to join the cluster.

**Note**: Cross account entries are not permitted for node type access entries.

**Note**: The IAM identity creating node type access entries of either `EC2_LINUX` or `EC2_WINDOWS` must have both iam:PassRole and ec2:RunInstances permissions. These permissions will be validated by EKS, and a failure will be returned if permissions are not present.

GitHub issues resolved:

- separate aws-auth for worker node and others [#822](https://github.com/aws/containers-roadmap/issues/822)
- add parameter at cluster creation for worker IAM role [#727](https://github.com/aws/containers-roadmap/issues/727)
- Managed Node Groups using role with extra path do not join cluster [#926](https://github.com/aws/containers-roadmap/issues/926)

### **Migration fully off the aws-auth ConfigMap**

The aws-auth ConfigMap can still be used until the to be announced future Kubernetes version where it no longer supported. But migration to access entry API is strongly encouraged.

**Migrate an entry in the aws-auth ConfigMap to access entry API**

For an existing cluster, change your cluster authenticationMode from `CONFIG_MAP` to `API_AND_CONFIG_MAP`. Before deleting anything in the aws-auth ConfigMap, create equivalent access entries. Specify the same username and/or groups already used the aws-auth ConfigMap if necessary. After access entry creation, then delete the corresponding entry from the ConfigMap and validate your enviroment. Once you have migrated everything to access entry APIs, you can switch your cluster authenticationMode to API, to turn off the aws-auth ConfigMap as a source for authenticated IAM principals.

### **AWS Service Integrations**

The access entries feature simplifies how other AWS services can obtain Kubernetes permissions needed to perform actions against EKS clusters. In the future, AWS services with EKS integrations (including EMR, Batch, Resilience Hub) will update their Service Linked Roles to contain permissions to create access entries. When you take an action in that service, the service SLR will automatically create an access entry and associate a specific EKS access policy for that service. There is no need for users to manually create access entries and associated permissions. For example, when Resilience Hub migrates to access entry, all of the steps currently outlined in this doc [page](https://docs.aws.amazon.com/resilience-hub/latest/userguide/enabling-eks-in-arh.html) will no longer be required. You can think of these "service-linked access entries" as the Kubernetes permission equivalent of [AWS service-linked-roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/using-service-linked-roles.html).

**IAM policy control for access entries**
 EKS access entry APIs support [condition context keys](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_condition-keys.html) to enable fine grained control over creation/management of access entries.

**Example policy a future service might add to its SLR to give itself permissions to create a specific access entry with only specific associated access policies.**
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "eks:CreateAccessEntry",
      "Resource": "arn:aws:eks:*:*:cluster/*",
      "Condition": {
        "ArnEquals": {
          "eks:principalArn": "${aws:principalArn}"
        },
        "StringEquals": {
          "aws:RequestTag/aws:eks:managed": "${aws:principalArn}"
        },
        "StringEqualsIfExists": {
           "eks:username": ""
        },
        "ForAllValues:StringEqualsIfExists": {
           "eks:kubernetesGroups": [ "" ]
        }
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": "eks:DeleteAccessEntry",
      "Resource": "arn:aws:eks:*:*:access-entry/*/role/*",
      "Condition": {
        "ArnEquals": {
          "eks:principalArn": "${aws:principalArn}"
        },
        "StringEquals": {
          "aws:ResourceTag/aws:eks:managed": "${aws:principalArn}"
      }
    },
    {
      "Effect": "Allow",
      "Action": "eks:AssociateAccessPolicy",
      "Resource": "arn:aws:eks:*:*:access-entry/*/role/*",
      "Condition": {
        "ArnEquals": {
          "eks:principalArn": "${aws:principalArn}"
        },
        "StringEquals": {
          "eks:policyArn": "arn:aws:eks::aws:cluster-access-policy/AmazonEKSNodeOperatorClusterPolicy",
          "eks:accessScope": "cluster",
          "aws:ResourceTag/aws:eks:managed": "${aws:principalArn}"
        }
      }
    },
    {
      "Effect": "Allow",
      "Action": "eks:AssociateAccessPolicy",
      "Resource": "arn:aws:eks:*:*:access-entry/*/role/*",
      "Condition": {
        "ArnEquals": {
          "eks:principalArn": "${aws:principalArn}"
        },
        "StringEquals": {
          "eks:policyArn": "arn:aws:eks::aws:cluster-access-policy/AmazonEKSNodeOperatorPolicy",
          "eks:accessScope": "namespace",
          "aws:ResourceTag/aws:eks:managed": "${aws:principalArn}"
        },
        "ForAllValues:StringEquals": {
          "eks:namespaces": [
            "my-namepsace" 
          ]
        }
      }
    }
  ]
}

```
You can use this example as a reference to write your own IAM policies that scope down what type of access entries IAM principals can create.

EKS supports the following new condition keys associated with this feature

- `eks:principalArn`
- `eks:username`
- `eks:kubernetesGroups`
- `eks:accessScope`
- `eks:namespaces`
- ` eks:policyArn`
- `eks:authenticationMode`

### **Deletion**

Deleting an access entry will remove any associated access policies. There is no need to first disassociate access policies.
```
aws eks delete-access-entry --cluster-name my-cluster \
   --principal-arn arn:aws:iam::012345678910:role/MyRole
```

### **SSO/AWS IAM Identity Center**

There is no direct change to improve this user experience. You still need to add each auto generated role as a separate access entry. But automating that process on your own should be become easier since everything is now an AWS API. We plan to improve this UX in the future by allowing users to directly associate IAM Identity Center [permission sets](https://docs.aws.amazon.com/singlesignon/latest/userguide/permissionsetsconcept.html) to EKS clusters. Follow this GitHub issue for details: EKS authentication rolearn wildcard support [#474](https://github.com/aws/containers-roadmap/issues/474).

**Comment by ab77** [comment]
There may be a bug in the CLoudFormation implementation of this feature, it seems that Acess Policies are lost on STACK_UPDATE (they are present after the initial create though). In the template, resources are defined as follows:

```
  AccessEntryAdmin:
    Type: AWS::EKS::AccessEntry
    Properties:
      ClusterName:
        Ref: Cluster
      PrincipalArn:
        Fn::Sub: arn:${AWS::Partition}:iam::${AWS::AccountId}:role/admin
      KubernetesGroups:
      - cluster-admin
      AccessPolicies:
      - PolicyArn:
          Fn::Sub: arn:${AWS::Partition}:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy
        AccessScope:
          Type: cluster
```

.. after update:

<img width="141" alt="image" src="https://github.com/aws/containers-roadmap/assets/2033996/e420bcc8-8470-4f9d-ba64-e1e8cb396503">

The only access policy that isn't wiped, is the one attached to the creator IAM role.

**Comment by lynnnnnnluo** [comment]
Hello,  thanks for your feedback ! I attempted to reproduce the issue in the following way but I wasn't able to reproduce 
I created Access entry with the template you have provided, then I triggered a stack update by updating the stack tags. Checking EKS console, the access policies are still there.

Could you describe more on the CFN template before/after the stack update so I can try reproducing ?

**Comment by ab77** [comment]
> Could you describe more on the CFN template before/after the stack update so I can try reproducing ?

Hello, I'll try to reproduce over here again (this is my cluster definition for now):

```
  Cluster:
    Type: AWS::EKS::Cluster
    Properties:
      Name: !Sub "${AWS::StackName}"
      RoleArn: !GetAtt Role.Arn
      AccessConfig:
        AuthenticationMode: API
        BootstrapClusterCreatorAdminPermissions: true
      ResourcesVpcConfig:
        EndpointPrivateAccess: true
        EndpointPublicAccess: true
        SecurityGroupIds:
          - Fn::ImportValue: !Sub "${ParentClientStack}-ClientSecurityGroup"
        SubnetIds:
          - Fn::ImportValue: !Sub "${ParentVPCStack}-SubnetAPrivate"
          - Fn::ImportValue: !Sub "${ParentVPCStack}-SubnetBPrivate"
      Version: !Ref Version
      EncryptionConfig:
        - Provider:
            KeyArn:
              Fn::ImportValue: !Sub "${ParentKmsKeyStack}-KeyArn"
          Resources:
            - secrets
      KubernetesNetworkConfig:
        IpFamily: !Ref IpFamily
```

.. also I looked through Cloudtrail logs and there are `DisassociateAccessPolicy` events from CFN, so it's definitely triggering disassociations for some reason on updates:

```
    "eventTime": "2023-12-26T20:36:16Z",
    "eventSource": "eks.amazonaws.com",
    "eventName": "DisassociateAccessPolicy",
    "awsRegion": "us-east-1",
    "sourceIPAddress": "cloudformation.amazonaws.com",
    "userAgent": "cloudformation.amazonaws.com",
    "requestParameters": {
        "policyArn": "arn%3Aaws%3Aeks%3A%3Aaws%3Acluster-access-policy%2FAmazonEKSClusterAdminPolicy",
        "name": "foo-bar-eks-1",
        "principalArn": "arn%3Aaws%3Aiam%3A%3A1234567890%3Arole%2Fadmin"
    },
```

**Comment by lynnnnnnluo** [comment]
Thanks for your reply!  It turns out there is a code bug in CFN update, we will be working on the fix as soon as possible.

**Comment by lynnnnnnluo** [comment]
Hello, the fix to above issue was shipped, thanks again for the feedback.

**Comment by joebowbeer** [comment]
@mikestef9 Do both the EKS User Guide and EKS Best Practices Guide need updating now that the Cluster Access Manager API has been added and is the preferred way to manage access of AWS IAM principals to Amazon EKS clusters?

https://github.com/aws/aws-eks-best-practices/issues/463

**Comment by bryantbiggs** [comment]
[`describe-access-entry`](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/eks/describe-access-entry.html) contains the `kubernetesGroups` property

**Comment by dennispan** [comment]
@mikestef9 Regarding SSO/IAM Identity Center
> You still need to add each auto generated role as a separate access entry. 

Where to get the auto generated role arn for federated users? The ARN from `aws sts get-caller-identity` is rejected when adding Access Entry with it.

**Comment by kennethredler** [comment]
> @mikestef9 Regarding SSO/IAM Identity Center
> > You still need to add each auto generated role as a separate access entry. 
> 
> Where to get the auto generated role arn for federated users? The ARN from `aws sts get-caller-identity` is rejected when adding Access Entry with it.

What principal type are you using (role, user, etc.)? What does the ARN look like which you are attempting to use with an access entry? What is the error message you see?

**Comment by dennispan** [comment]
@kennethredler This is a role from sso, e.g. `arn:aws:sts::012345678901:assumed-role/AWSReservedSSO_UserAccess_d2506a67da7d7d9b/dennis.pan@company.com`

This returns error:
```
The principalArn parameter format is not valid
```

I've also tried changing its format, which works with `aws-auth` config map, e.g. `arn:aws:iam::012345678901:role/AWSReservedSSO_UserAccess_d2506a67da7d7d9b`

That gives this error:
```
The specified principalArn is invalid: invalid principal.
```

**Comment by MartinEmrich** [comment]
Try removing the `role/`. This snippet is from a working eksctl ClusterConfig:
````
accessConfig:
  authenticationMode: API_AND_CONFIG_MAP
  accessEntries:
    - principalARN: "arn:aws:iam::123456789012:AWSReservedSSO_MyAdministrators_123456789abcdef"
      type: STANDARD
      kubernetesGroups:
        - system:masters
````

**Comment by dennispan** [comment]
Still got `The principalArn parameter format is not valid`

**Comment by bryantbiggs** [comment]
if you use Terraform, you can use the `aws_iam_roles` data source which will accept a regex and then pass the first match to the access entry. This helps work around the random bits at the end of the generated identity store name:

```hcl
data "aws_iam_roles" "admin" {
  name_regex  = "AWSReservedSSO_MyAdministrators_.*"
  path_prefix = "/aws-reserved/sso.amazonaws.com/"
}

module "eks" {
  source  = "terraform-aws-modules/eks/aws"
  version = "~> 20.24"

  # Truncated for brevity ...

  authentication_mode = "API"

  access_entries = {
    operators = {
      principal_arn = tolist(data.aws_iam_roles.admin.arns)[0]

      policy_associations = {
        operators = {
          policy_arn = "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"
          access_scope = {
            type = "cluster"
          }
        }
      }
    }
  }
}
```

**Comment by bryantbiggs** [comment]
> kubernetesGroups:
>         - system:masters

also, this is not part is not valid 

`system:masters` is not supported in cluster access entry. instead, you would associate your role to the `"arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"` policy

**Comment by MartinEmrich** [comment]
@bryantbiggs what makes you say that? This works fine here (EKS 1.29.5). 
Docs for the `kubernetesGroups:` option: https://eksctl.io/usage/access-entries/

Analyzing the current state, indeed EKS mapped system:masters to `"policyArn": "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"`, so it effectively had the same effect.
If this is just a temporary compatibility layer, please let me/us know.

OFC giving users full admin privileges it is usually not the most secure/best-practisey thing to do, but this is a playground cluster for developers, so no worries there.

**Comment by cdenneen** [comment]
> @bryantbiggs what makes you say that? This works fine here (EKS 1.29.5). Docs for the `kubernetesGroups:` option: https://eksctl.io/usage/access-entries/
> 
> Analyzing the current state, indeed EKS mapped system:masters to `"policyArn": "arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy"`, so it effectively had the same effect. If this is just a temporary compatibility layer, please let me/us know.
> 
> OFC giving users full admin privileges it is usually not the most secure/best-practisey thing to do, but this is a playground cluster for developers, so no worries there.

@MartinEmrich because you are mixing ConfigMap and Access entries:
https://eksctl.io/usage/access-entries/#iam-entities

So from the example in the docs:

```yaml
accessConfig:
  authenticationMode: API_AND_CONFIG_MAP
  accessEntries:
    - principalARN: arn:aws:iam::111122223333:user/my-user-name
      type: STANDARD
      kubernetesGroups: # optional Kubernetes groups
        - group1 # groups can used to give permissions via RBAC
        - group2

    - principalARN: arn:aws:iam::111122223333:role/role-name-1
      accessPolicies: # optional access polices
        - policyARN: arn:aws:eks::aws:cluster-access-policy/AmazonEKSViewPolicy
          accessScope:
            type: namespace
            namespaces:
              - default
              - my-namespace
              - dev-*

    - principalARN: arn:aws:iam::111122223333:role/admin-role
      accessPolicies: # optional access polices
        - policyARN: arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy
          accessScope:
            type: cluster

    - principalARN: arn:aws:iam::111122223333:role/role-name-2
      type: EC2_LINUX
```
the `type: STANDARD` with `kubernetesGroups` would be for the config map and the following would be for an access entry (no kubernetesGroups because it's not creating an entry in the configmap so doesn't need to map a kubernetes group like `system:masters`, instead it has `accessPolicies` and `accessScope`):

```
    - principalARN: arn:aws:iam::111122223333:role/admin-role
      accessPolicies: # optional access polices
        - policyARN: arn:aws:eks::aws:cluster-access-policy/AmazonEKSClusterAdminPolicy
          accessScope:
            type: cluster
```

@bryantbiggs clarify if I'm wrong

**Comment by cdenneen** [comment]
> Try removing the `role/`. This snippet is from a working eksctl ClusterConfig:
> 
> ```
> accessConfig:
>   authenticationMode: API_AND_CONFIG_MAP
>   accessEntries:
>     - principalARN: "arn:aws:iam::123456789012:AWSReservedSSO_MyAdministrators_123456789abcdef"
>       type: STANDARD
>       kubernetesGroups:
>         - system:masters
> ```

so this example you are creating a configMap entry vs an accessScope. Which is using the `CONFIG_MAP` part of the authentication instead of the `API` part. The `API` part will deprecate the `CONFIG_MAP` part in the future I believe on the road map but not sure when. 

I think the confusion here is the `eksctl` documentation is using the term `accessEntries` which should be used solely for `API` authentication so since all `CONFIG_MAP` and `API` (accessEntries) are created under `accessEntries` in the eksctl configuration I can see where the confusion might happen here.

**Comment by bryantbiggs** [comment]
@cdenneen that is correct! the use of `system:*` within the `aws-auth` ConfigMap would be replaced with the provided EKS cluster access policy equivalents when using EKS cluster access entries

**Comment by dennispan** [comment]
Thanks @joshuabaird ! `arn:aws:iam::1234455435:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_Blah_834943dcf9f248` does work. How should I understand `role/aws-reserved/sso.amazonaws.com/`? I mean is that something that we get from `sts`?

**Comment by MartinEmrich** [comment]
@bryantbiggs @cdenneen Thanks! 

So my eksctl ClusterConfig basically created both entries (AccessEntry on the "outside", `aws-auth` ConfigMap entry on the "inside"), and they have in "hybrid" API_AND_CONFIG_MAP mode just the same effect?

Will play around with pure `API` mode in the future and see what happens...

**Comment by atheiman** [comment]
@bryantbiggs - Im running into an issue where an identity in an EKS has the `AmazonEKSAdminPolicy` access policy granted, but cannot create a Role + RoleBinding which declares permissions that exist in the access policy? The error is `attempting to grant rbac permissions not currently held`. Are the permissions contained in an eks access policy not inspected by kubernetes when determining if the identity is allowed to pass permissions to another subject?

Edit - I have confirmed that access policies less than `AmazonEKSClusterAdminPolicy` are limited to not capable of creating (Cluster)Role / (Cluster)RoleBinding resources b/c they do not include `escalate` or `bind` verbs _and_ their rules are not visible to kubernetes internals when it is evaluating if the user has at least the permissions declared in the Role / RoleBinding request.
